include/dali/imgcodec/image_codec.h

```diff
@@ -0,0 +1,153 @@
+
+#include <memory>
+#include <stdexcept>
+#include <vector>
+#include "dali/core/any.h"
+#include "dali/core/span.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/imgcodec/image_format.h"
+#include "dali/pipeline/data/sample_view.h"
+#include "dali/pipeline/data/backend.h"
+
+namespace dali {
+class ThreadPool;
+
+namespace imgcodec {
+template <typename T, span_extent_t E = dynamic_extent>
+using cspan = span<const T, E>;
+
+struct DecodeParams {
+  bool use_roi;
+  struct {
+    TensorShape<> begin, end;
+  } roi;
+  DALIDataType dtype;
+  DALIImageType format;
+  bool planar;  // not required initially
+};
+
+struct DecodeResult {
+  bool success;
+  std::exception_ptr exception;
+};
+
+struct ImageCodecProperties {
+  /**
+   * @brief Whether the codec can decode a part of the image without storing the
+   *        entire image in memory.
+   */
+  bool roi_support = false;
+
+  /**
+   * @brief A mask of supported input kinds
+   */
+  InputKind supported_input_kinds;
+
+  // if true and the codec fails to decode an image,
+  // an attempt will be made to use other compatible codecs
+  bool fallback = true;
+};
+
+class DLL_PUBLIC ImageCodecInstance {
+ public:
+  virtual ~ImageCodecInstance() = default;
+  /**
+   * @brief Checks whether this codec can decode this encoded image with given parameters
+   */
+  virtual bool CanDecode(ImageSource *in, DecodeParams opts) = 0;
+
+  /**
+   * @brief Batch version of CanDecode
+   */
+  virtual std::vector<bool> CanDecode(cspan<ImageSource *> in, cspan<DecodeParams> opts) = 0;
+
+  /**
+   * @brief Decodes a single image to a host buffer
+   */
+  virtual DecodeResult Decode(SampleView<CPUBackend> out, ImageSource *in, DecodeParams opts) = 0;
+
+  /**
+   * @brief Decodes a batch of images to host buffers
+   */
+  virtual std::vector<DecodeResult> Decode(span<SampleView<CPUBackend>> out,
+                                           cspan<ImageSource *> in, cspan<DecodeParams> opts) = 0;
+
+
+
+  /**
+   * @brief Decodes a single image to a device buffer
+   */
+  virtual DecodeResult Decode(SampleView<GPUBackend> out, ImageSource *in, DecodeParams opts) = 0;
+
+  /**
+   * @brief Decodes a single image to device buffers
+   */
+  virtual std::vector<DecodeResult> Decode(span<SampleView<GPUBackend>> out,
+                                           cspan<ImageSource *> in, cspan<DecodeParams> opts) = 0;
+
+  /**
+   * @brief Sets a codec-specific parameter
+   */
+  virtual void SetParam(const char *key, const any &value) = 0;
+  /**
+   * @brief Gets a codec-specific parameter
+   */
+  virtual any GetParam(const char *key) const = 0;
+
+  template <typename T>
+  inline enable_if_t<!std::is_same<std::remove_reference_t<T>, any>::value>
+  SetParam(const char *key, T value) {
+    SetParam(key, any(value));
+  }
+
+  template <typename T>
+  inline T GetParam(const char *key) const {
+    return any_cast<T>(GetParam(key));
+  }
+};
+
+class DLL_PUBLIC ImageCodec {
```

Comment:
Could you explain how these will be created and kept during the pipeline or program lifetime? Is there any registry or something?

Reply:
Yes, see ImageFormatRegistry. You register formats and you register codecs for those formats. The `ImageCodec` is used as a factory to create `ImageCodecInstance`. There will likely be like a codec instance cache on a higher level abstraction that will be used by the ops. This will be part of the next PR.


```
// Note: We will have a single format registry for DALI 
ImageFormatRegistry registry;
ImageFormat jpeg("jpeg", std::make_shared<JpegParser>());
jpeg.RegisterCodec(std::make_shared<JpegCodec>(), 0.0f);


// Inside a particular OP
auto format = registry.GetImageFormat(image_source);

for (auto codec : format.Codecs()) {
   if (...) {  // for example limiting to host mem codecs
     if (codec instance in cache) {
         // use codec instance
     } else {
         codec_instance = codec.Create(...)        
         // cache codec instance
         // use codec instance   
     }
   }
}

```

Reply:
Yeah, I was asking about where `ImageCodec` comes from.